### PR TITLE
[AMDGPU] Extend permlane16, permlanex16 and permlane64 intrinsic lowering for generic types

### PR DESCRIPTION
**Pull request by vikramRH**
Kindly review only top commits here (i.e commits except the first). These are incremental changes over #89217 , with core logic being the same.  Only reason to split these up into separate PR is for ease of review.
This patch along with #89217 and #91190 should get us ready to enable 64 bit optimizations in atomic optimizer.

### CODE REVIEW

File: clang/lib/CodeGen/CGBuiltin.cpp [hunk 1]
```diff
@@ -18479,6 +18479,25 @@ Value *CodeGenFunction::EmitAMDGPUBuiltinExpr(unsigned BuiltinID,
         CGM.getIntrinsic(Intrinsic::amdgcn_update_dpp, Args[0]->getType());
     return Builder.CreateCall(F, Args);
   }
+  case AMDGPU::BI__builtin_amdgcn_permlane16:
+  case AMDGPU::BI__builtin_amdgcn_permlanex16: {
+    Intrinsic::ID IID;
+    IID = BuiltinID == AMDGPU::BI__builtin_amdgcn_permlane16
```
Comment (arsenm) [inline clang/lib/CodeGen/CGBuiltin.cpp:18485]: combine declare + define, also can sink down to use

File: clang/lib/CodeGen/CGBuiltin.cpp [hunk 2]
```diff
@@ -18479,6 +18479,25 @@ Value *CodeGenFunction::EmitAMDGPUBuiltinExpr(unsigned BuiltinID,
...
+              ? Intrinsic::amdgcn_permlane16
+              : Intrinsic::amdgcn_permlanex16;
+
+    llvm::Value *Src0 = EmitScalarExpr(E->getArg(0));
+    llvm::Value *Src1 = EmitScalarExpr(E->getArg(1));
+    llvm::Value *Src2 = EmitScalarExpr(E->getArg(2));
```
Comment (arsenm) [inline clang/lib/CodeGen/CGBuiltin.cpp:18491]: I assume EmitScalarExpr handles the immargs correctly?

File: llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp [hunk 3]
```diff
@@ -5433,7 +5450,16 @@ bool AMDGPULegalizerInfo::legalizeLaneOp(LegalizerHelper &Helper,
                             ? Src0
                             : B.buildBitcast(LLT::scalar(Size), Src0).getReg(0);
     Src0 = B.buildAnyExt(S32, Src0Cast).getReg(0);
-    if (Src2.isValid()) {
+
+    if (IsPermLane16) {
+      Register Src1Cast =
+          MRI.getType(Src1).isScalar()
+              ? Src1
+              : B.buildBitcast(LLT::scalar(Size), Src2).getReg(0);
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp:5458]: Like the other patch, shouldn't need any bitcasts

**Review by arsenm — COMMENTED** [review]
On this and the previous, can you add a section to AMDGPUUsage for the intrinsics and what types they support

File: clang/lib/CodeGen/CGBuiltin.cpp (hunk 2 above)
Comment (vikramRH) [reply clang/lib/CodeGen/CGBuiltin.cpp:18486]: yes

File: llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp (hunk 3 above)
Comment (vikramRH) [reply llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp:5458]: Yes, I will take over the changes from https://github.com/llvm/llvm-project/pull/89217 once finalized,

**Comment by vikramRH** [comment]
1. Added/updated tests for permlanex16, permlane64
2. This needs https://github.com/llvm/llvm-project/pull/89217 to land first so that only incremental changes can be reviewed.

File: clang/lib/CodeGen/CGBuiltin.cpp [hunk 4]
```diff
@@ -18479,6 +18479,28 @@ Value *CodeGenFunction::EmitAMDGPUBuiltinExpr(unsigned BuiltinID,
         CGM.getIntrinsic(Intrinsic::amdgcn_update_dpp, Args[0]->getType());
     return Builder.CreateCall(F, Args);
   }
+  case AMDGPU::BI__builtin_amdgcn_permlane16:
+  case AMDGPU::BI__builtin_amdgcn_permlanex16: {
+    llvm::Value *Src0 = EmitScalarExpr(E->getArg(0));
```
Comment (arsenm) [inline clang/lib/CodeGen/CGBuiltin.cpp:18484]: If there's really not a helper to just EmitScalarExpr for N arguments, there should be one used here

Comment (vikramRH) [reply clang/lib/CodeGen/CGBuiltin.cpp]: added a new helper

**Comment by vikramRH** [comment]
Updated this PR to be in sync with #89217, However still plan is to land this land this only after changes in #89217 are accepted.

**Review by arsenm — APPROVED** [review]

**Pull request merged**